lib/SIL/SILFunctionType.cpp

```diff
@@ -148,22 +150,68 @@ CanSILFunctionType SILFunctionType::getWithoutDifferentiability() {
                               getOptionalErrorResult(), getASTContext());
 }
 
+// Returns the canonical generic signature for an autodiff associated function
+// given an existing associated function generic signature. All differentiation
+// parameters are constrained to conform to `Differentiable`.
+static CanGenericSignature getAssociatedFunctionGenericSignature(
```

Comment:
I thought about unifying `getAssociatedFunctionGenericSignature` defined here in `lib/SIL/SILFunctionType.cpp` (1) with the one defined in `lib/SILOptimizer/Differentiation.cpp` (2), but the two are not quite compatible.

(1) takes a pre-existing associated function generic signature, but (2) builds the associated function generic signature given the original function's generic signature and `[differentiable]` attribute requirements.

The part that can be shared is the "constrain all wrt parameters to conform to `Differentiable`" logic - I don't sharing that short code is worthwhile.